test/parallel-benchmark/mzcompose.py

```diff
@@ -0,0 +1,1190 @@
+    def generate(self, duration: int) -> Iterator[float]:
+        raise NotImplementedError
+
+
+class Periodic(Distribution):
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[float]:
+        next_time = time.time()
+        for i in range(duration * self.per_second):
+            yield next_time
+            next_time += 1 / self.per_second
+            sleep_until(next_time)
+
+
+class Gaussian(Distribution):
+    """Run the action with a sleep time between actions drawn from a Gaussian distribution"""
+
+    def __init__(self, mean: float, stddev: float):
+        self.mean = mean
+        self.stddev = stddev
+
+    def generate(self, duration: int) -> Iterator[float]:
+        end_time = time.time() + duration
+        next_time = time.time()
+        while time.time() < end_time:
+            yield next_time
+            next_time += max(0, random.gauss(self.mean, self.stddev))
+            sleep_until(next_time)
+
+
+class PhaseAction:
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        raise NotImplementedError
+
+
+class OpenLoop(PhaseAction):
+    def __init__(self, action: Action, dist: Distribution):
+        self.action = action
+        self.dist = dist
+
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        for start_time in self.dist.generate(duration):
+            jobs.put(lambda: self.action.run(start_time, conns, measurements))
+
+
+class ClosedLoop(PhaseAction):
+    def __init__(self, action: Action):
+        self.action = action
+
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        end_time = time.time() + duration
+        while time.time() < end_time:
+            self.action.run(time.time(), conns, measurements)
+
+
+class Phase:
+    def run(
+        self,
+        c: Composition,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        raise NotImplementedError
+
+
+class TdPhase(Phase):
+    def __init__(self, td: str):
+        self.td = dedent(td)
+
+    def run(
+        self,
+        c: Composition,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        c.testdrive(self.td, quiet=True)
+
+
+class LoadPhase(Phase):
+    duration: int
+    phase_actions: list[ClosedLoop | OpenLoop]
```

Comment:
```suggestion
    phase_actions: list[PhaseAction]
```

Reply:
mypy is a bit annoying here:
```
  /home/deen/git/materialize2/test/parallel-benchmark/mzcompose.py:415:29 - error: Argument of type "list[ClosedLoop | OpenLoop]" cannot be assigned to parameter "actions" of type "list[PhaseAction]" in function "__init__"
```
I don't want to list the type everywhere explicitly.

Reply:
Ah, ok. You could avoid this issue by using a sequence instead of a list.